homeassistant/components/teslemetry/__init__.py

```diff
@@ -52,36 +57,56 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     energysites: list[TeslemetryEnergyData] = []
     for product in products:
         if "vin" in product and Scope.VEHICLE_DEVICE_DATA in scopes:
+            # Remove the protobuff 'cached_data' that we do not use to save memory
+            product.pop("cached_data", None)
             vin = product["vin"]
             api = VehicleSpecific(teslemetry.vehicle, vin)
-            coordinator = TeslemetryVehicleDataCoordinator(hass, api)
+            coordinator = TeslemetryVehicleDataCoordinator(hass, api, product)
+            device = DeviceInfo(
+                identifiers={(DOMAIN, vin)},
+                manufacturer="Tesla",
+                configuration_url="https://teslemetry.com/console",
+                name=product["display_name"],
+                model=MODELS.get(vin[3]),
+                serial_number=vin,
+            )
+
             vehicles.append(
                 TeslemetryVehicleData(
                     api=api,
                     coordinator=coordinator,
                     vin=vin,
+                    device=device,
                 )
             )
         elif "energy_site_id" in product and Scope.ENERGY_DEVICE_DATA in scopes:
             site_id = product["energy_site_id"]
             api = EnergySpecific(teslemetry.energy, site_id)
+            live_coordinator = TeslemetryEnergySiteLiveCoordinator(hass, api)
+            device = DeviceInfo(
+                identifiers={(DOMAIN, str(site_id))},
+                manufacturer="Tesla",
+                configuration_url="https://teslemetry.com/console",
+                name=product.get("site_name", "Energy Site"),
+            )
+
             energysites.append(
                 TeslemetryEnergyData(
                     api=api,
-                    coordinator=TeslemetryEnergyDataCoordinator(hass, api),
+                    live_coordinator=live_coordinator,
                     id=site_id,
-                    info=product,
+                    device=device,
                 )
             )
 
-    # Do all coordinator first refreshes simultaneously
+    # Run all first refreshes
     await asyncio.gather(
         *(
             vehicle.coordinator.async_config_entry_first_refresh()
             for vehicle in vehicles
         ),
         *(
-            energysite.coordinator.async_config_entry_first_refresh()
+            energysite.live_coordinator.async_config_entry_first_refresh()
```

Comment:
On line 115 (out of reach for this review) in a follow-up move from `hass.data` to `entry.runtime_data`